### PR TITLE
Seller services controller

### PR DESCRIPTION
**Pull request by tavisca-acharpe**

### CODE REVIEW

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 1]
```diff
@@ -23,7 +24,7 @@ public class Product
         public Nullable<DateTime> PurchasedDate { get; set; }
         public Address PickupAddress { get; set; }
 
-        IProductStore productStore;
+        IProductStore productStore = new MockProductStore();
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: core models should be unaware of your implementations

File: src/OnlineRetailPortal.Services/Services/ProductService.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using System.Threading.Tasks;
+using OnlineRetailPortal.Mock;
+
+namespace OnlineRetailPortal.Services
+{
+    public class ProductService : IProductService
+    {
+        IProductStoreFactory productStoreFactory = new Mock.ProductStoreFactory();
+        IProductStore productStore ;
+        Core.Product product;
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Services/Services/ProductService.cs:14]: why do we need this at root level? also the dependencies should be injected in a constructor

File: src/OnlineRetailPortal.Services/Translators/AddProductRequestTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                Description = addProductRequest.Description,
+                HeroImage = new Image { Url = addProductRequest.HeroImage.Url },
+                Price = new Price { Amount = addProductRequest.Price.Amount, IsNegotiable = addProductRequest.Price.IsNegotiable },
+                Category = (Category)addProductRequest.Category,
+                PurchasedDate = addProductRequest.PurchasedDate,
+                Images = addProductRequest.Images.Select(x => new Image
+                {
+                    Url = x.Url
+                }).ToList(),
+                PickupAddress = new Address
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Services/Translators/AddProductRequestTranslator.cs]: please add null checks for optional fields

File: src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                HeroImage = new Contracts.Image { Url = addProductRequest.HeroImage.Url },
+                Price = new Contracts.Price { Amount = addProductRequest.Price.Amount, IsNegotiable = addProductRequest.Price.IsNegotiable },
+                Category = (Contracts.Category)addProductRequest.Category,
+                Images = new Contracts.Product().Images = addProductRequest.Images.Select(x => new Contracts.Image
+                {
+                    Url = x.Url
+                }).ToList(),
+                PurchasedDate = addProductRequest.PurchasedDate,
+                PickupAddress = new Contracts.Product().PickupAddress = new Contracts.Address()
+                {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs]: please add null checks for optional fields

File: src/OnlineRetailPortal.Web/Validations/Validator.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using FluentValidation.Results;
+using FluentValidation;
+using System.Linq;
+
+namespace OnlineRetailPortal.Web.Validations
+{
+    public static class Validator
+    {
+        public static void EnsureValid<AddProductRequest>(this AbstractValidator<AddProductRequest> validator, AddProductRequest request)
+        {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Validations/Validator.cs]: skipping the review since you are still working on this

**Review by Tavisca-smohite — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 1 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: done

File: src/OnlineRetailPortal.Services/Translators/AddProductRequestTranslator.cs (hunk 3 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Services/Translators/AddProductRequestTranslator.cs]: done

File: src/OnlineRetailPortal.Core/ErrorCodes.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+using OnlineRetailPortal.Contracts;
+
+
+namespace OnlineRetailPortal.Core
+{
+    public class ErrorCodes
+    {
+        public static string Invalid()
+        {
+            return string.Format(ErrorCodesMessage.Invalid);
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Core/ErrorCodes.cs]: Rename ErrorCodeMessage to ErrorMessage

File: src/OnlineRetailPortal.Web/Controllers/ProductsController.cs [hunk 7]
```diff
@@ -5,12 +5,31 @@
...
 {
     [Route("api/v1.0/onlineretailportal")]
     [ApiController]
     public class ProductsController : ControllerBase
     {
+        private readonly IProductService _productService;
+
+        public ProductsController(IProductService productService)
+        {
+            this._productService = productService;
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Controllers/ProductsController.cs]: Do we need 'this.'?

File: src/OnlineRetailPortal.Web/Models/Category.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        Properties,
+        Cars,
+        Furniture,
+        Mobiles,
+        Bikes,
+        Books,
+        Fashions,
+        Electronics,
+        Others
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Models/Category.cs]: These need be singular.

File: src/OnlineRetailPortal.Web/Startup.cs [hunk 9]
```diff
@@ -26,6 +28,12 @@ public Startup(IConfiguration configuration)
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddControllers();
+            services.AddSingleton<IProductService>(new Services.ProductService());
+        }
+
+        private object ProductService(IServiceProvider arg)
+        {
+            throw new NotImplementedException();
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Startup.cs]: Do we need this?

File: src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        public static Contracts.AddProductRequest ToDataContract(this AddProductRequest addProductRequest)
+        {
+            Contracts.AddProductRequest request = new Contracts.AddProductRequest()
+            {
+                SellerId = addProductRequest.SellerId,
+                Name = addProductRequest.Name,
+                Description = addProductRequest.Description,
+                HeroImage = new Contracts.Image { Url = addProductRequest.HeroImage.Url },
+                Price = new Contracts.Price { Value = new Contracts.Value(addProductRequest.Price.Value.Amount, addProductRequest.Price.Value.Currency), IsNegotiable = addProductRequest.Price.IsNegotiable },
+                Category = (Contracts.Category)addProductRequest.Category,
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs]: Use translators with ToEntity/ToModel instead of casting.

Comment (tavisca-swapnil) [reply src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs]: Comment applies throughtout.

File: src/OnlineRetailPortal.Web/Validations/AddProductRequestValidator.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            .WithMessage(Error.NullField("Description"))
+            .NotEmpty()
+            .WithErrorCode(ErrorCodes.MissingField())
+            .WithMessage(Error.MissingField("Description"))
+            .Length(4, 100)
+            .WithErrorCode(ErrorCodes.GreaterCharacter())
+            .WithMessage(Error.GreaterCharacter("Description", "4"));
+
+            RuleFor(x => x.PurchasedDate)
+            .LessThan(DateTime.Today)
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Validations/AddProductRequestValidator.cs:73]: Add Error Code

File: src/OnlineRetailPortal.Web/Validations/EnsureValidator.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            var validationResult = validator.Validate(request);
+
+            if (request == null)
+                throw new Exception(Error.MissingField("Product"));
+                    
+            if (validationResult.IsValid == false)
+            {
+                foreach (var error in validationResult.Errors)
+                {
+                    info.Add(new ErrorInfo { Code = int.Parse(error.ErrorCode), Message = error.ErrorMessage });
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Validations/EnsureValidator.cs]: Keep 'Code' as a string.

**Review by tavisca-swapnil — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Core/ErrorCodes.cs (hunk 6 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Core/ErrorCodes.cs]: done

File: src/OnlineRetailPortal.Web/Models/Category.cs (hunk 8 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Models/Category.cs]: done

File: src/OnlineRetailPortal.Web/Startup.cs (hunk 9 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Startup.cs]: no removed

File: src/OnlineRetailPortal.Web/Validations/AddProductRequestValidator.cs (hunk 11 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Validations/AddProductRequestValidator.cs:73]: done

File: src/OnlineRetailPortal.Web/Validations/EnsureValidator.cs (hunk 12 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Validations/EnsureValidator.cs]: done

File: src/OnlineRetailPortal.Web/Controllers/ProductsController.cs (hunk 7 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Controllers/ProductsController.cs]: no

File: src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs (hunk 10 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Web/Translators/AddProductRequestTranslator.cs]: done

File: src/OnlineRetailPortal.Services/Services/ProductService.cs (hunk 2 above)
Comment (tavisca-acharpe) [reply src/OnlineRetailPortal.Services/Services/ProductService.cs:14]: added only to test

**Comment by tavisca-acharpe** [comment]
done in dif branch

**Pull request closed without merging**